drivers/mspi/mspi_esp32.h

```diff
@@ -0,0 +1,59 @@
+/*
+ * Copyright (c) 2025 Pavel Maloletkov.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr/drivers/mspi.h>
+#include <zephyr/drivers/clock_control.h>
+
+#include <hal/spi_hal.h>
+
+#ifdef SOC_GDMA_SUPPORTED
+#include <hal/gdma_hal.h>
+#include <hal/gdma_ll.h>
+#endif
+
+#define MSPI_MAX_FREQ             80000000
+#define MSPI_TIMEOUT_MS           100
+#define SPI_DUTY_CYCLE_50_PERCENT 128
+
+/* Driver configuration structure */
+struct mspi_esp32_config {
+	spi_dev_t *spi;
+	const struct pinctrl_dev_config *pcfg;
+	const struct device *clock_dev;
+	clock_control_subsys_t clock_subsys;
+	struct mspi_cfg mspi_config;
+	uint32_t peripheral_id;
+	uint32_t clock_frequency;
+	soc_module_clk_t clock_source;
+	bool dma_enabled;
+	int dma_clk_src;
+	int dma_host;
+	int max_dma_buf_size;
+	bool line_idle_low;
+	bool use_iomux;
+	uint32_t duty_cycle;
+	uint32_t input_delay_ns;
+};
+
+/* Driver data structure */
+struct mspi_esp32_data {
+	spi_hal_context_t hal_ctx;
+	spi_hal_config_t hal_config;
+	spi_hal_dev_config_t hal_dev_config;
+	struct mspi_dev_cfg mspi_dev_config;
+	struct k_mutex lock;
+	mspi_callback_handler_t callback;
+	void *callback_ctx;
+	uint32_t callback_mask;
+	uint32_t clock_source_hz;
```

Comment:
Hmm is this runtime changeable? If not it should go to the config. 

The data structure is aimed for driver runtime context, e.g: data that may change depending on the driver operating state.

Reply:
Technically yes, it's not supposed to change, but it's not known until `esp_clk_tree_src_get_freq_hz` is called during the configuration. I think there is no other way to retrieve it at the time of the driver instantiation.